tests/module_1/test_meteo_api.py

```diff
@@ -1,7 +1,116 @@
-""" This is a dummy example to show how to import code from src/ for testing"""
+import pandas as pd
+import requests
+import pytest
+from pytest import MonkeyPatch, LogCaptureFixture
+from unittest.mock import Mock
 
-from src.module_1.module_1_meteo_api import main
+from src.module_1.module_1_meteo_api import compute_yearly_mean_and_std, fetch_url
+
+
+def test_compute_yearly_mean_and_std() -> None:
+    test_parameter = "temperature"
+
+    data = {
+        "daily": {
+            "time": ["1950-01-01", "1950-01-02", "1951-01-01", "1951-01-02"],
+            f"{test_parameter}_model1": [1, 2, 3, 4],
+            f"{test_parameter}_model2": [30, 80, 50, 40],
+            f"{test_parameter}_model3": [10, 18, 17, 19],
+        }
+    }
+
+    expected = pd.DataFrame(
+        {
+            ("year", ""): {0: 1950, 1: 1951},
+            (f"{test_parameter}_model1", "mean"): {0: 1.5, 1: 3.5},
+            (f"{test_parameter}_model1", "std"): {
+                0: 0.7071067811865476,
+                1: 0.7071067811865476,
+            },
+            (f"{test_parameter}_model2", "mean"): {0: 55.0, 1: 45.0},
+            (f"{test_parameter}_model2", "std"): {
+                0: 35.35533905932738,
+                1: 7.0710678118654755,
+            },
+            (f"{test_parameter}_model3", "mean"): {0: 14.0, 1: 18.0},
+            (f"{test_parameter}_model3", "std"): {
+                0: 5.656854249492381,
+                1: 1.4142135623730951,
+            },
+        }
+    )
+
+    pd.testing.assert_frame_equal(
+        compute_yearly_mean_and_std(data), expected, check_dtype=False
+    )
+
+
+class MockResponse:
+    def __init__(
+        self, json_data: str, status_code: int, reason: str, headers: dict = {}
+    ) -> None:
+        self.json_data = json_data
+        self.status_code = status_code
+        self.reason = reason
+        self.headers = headers
+
+    def raise_for_status(self) -> None:
+        if self.status_code != 200:
+            raise requests.exceptions.RequestException(response=self)
+
+    def json(self) -> str:
+        return self.json_data
+
+
+def test_fetch_url_200(monkeypatch: MonkeyPatch) -> None:
+    mock_response = Mock(
+        return_value=MockResponse("mocked_response", 200, "mocked_reason")
+    )
+    monkeypatch.setattr("requests.get", mock_response)
+
+    response = fetch_url("mock_url")
+
+    assert response == "mocked_response"
```

Comment:
Perhaps overkill in this case, but in a real test check also the status code and the reason. If they are not relevant in your test, then the ```MockResponse``` class needs default values for them.

Reply:
Yeah, I thought about doing that but my `fetch_url` function returns directly the `response.json()` data instead of the response itself, so I cannot access the `response.status_code`.

Does this mean that in terms of design, it would be better for the `fetch_url` function to return the `response` instead of `response.json()`, thus, it is easier to test?

Reply:
Depends a lot on the context, but generally speaking yes. Statuses can tell you what flow to trigger for example, so they might be useful. 

In my case I like creating a dataclass or similar to load the info from response, with status being a field. Then, you don't answer back with a general response nor a native dict, but a custom class following the schema I want.